ethcore/src/client/client.rs

```diff
@@ -1652,6 +1652,33 @@ impl MiningBlockChainClient for Client {
 		open_block
 	}
 
+	fn reopen_block(&self, block: ClosedBlock) -> OpenBlock {
+		let engine = &*self.engine;
+		let mut block = block.reopen(engine);
+		let max_uncles = engine.maximum_uncle_count();
+		if block.uncles().len() < max_uncles {
+			let chain = self.chain.read();
+			let h = chain.best_block_hash();
+			// Add new uncles
+			let uncles = chain
+				.find_uncle_hashes(&h, engine.maximum_uncle_age())
+				.unwrap_or_else(Vec::new);
+
+			for h in uncles {
```

Comment:
I like the uncle-adding code in `prepare_open_block` more. Seems that the only difference is a check that `uncle` is already in the block, can't we use `filter` to get rid of such uncles in the list returned by `find_uncle_hashes` and re-use the same code in both methods?